src/stdx/mlock.zig

```diff
@@ -0,0 +1,112 @@
+            // would be difficult to gather for non-heap memory that is also faulted in,
+            // such as the stack, globals, etc.
+        },
+        .windows => try memory_lock_allocated_windows(options.allocated_size),
+        else => @compileError("unsupported platform"),
+    }
+}
+
+fn memory_lock_allocated_linux() MemoryLockError!void {
+    const MCL_CURRENT = 1; // Lock all currently mapped pages.
+    const MCL_ONFAULT = 3; // Lock all pages faulted in (i.e. stack space).
+    const result = os.linux.syscall1(.mlockall, MCL_CURRENT | MCL_ONFAULT);
+    switch (os.linux.E.init(result)) {
+        .SUCCESS => return,
+        .AGAIN => log.warn(mlockall_error, .{"some addresses could not be locked"}),
+        .NOMEM => log.warn(mlockall_error, .{"memory would exceed RLIMIT_MEMLOCK"}),
+        .PERM => log.warn(mlockall_error, .{
+            "insufficient privileges to lock memory",
+        }),
+        .INVAL => unreachable, // MCL_ONFAULT specified without MCL_CURRENT.
+        else => |err| return stdx.unexpected_errno("mlockall", err),
+    }
+    return .memory_not_locked;
+}
+
+fn memory_lock_allocated_windows(allocated_size: usize) MemoryLockError!void {
+    // Windows has VirtualLock which works similar to mlock with an address range.
+    // It would be difficult to gather the addresses of non-heap memory that is also
+    // faulted in, such as the stack, globals, etc. SetProcessWorkingSetSize can be
+    // used instead to lock all existing pages into memory to avoid swapping.
+    const set_process_working_set_size = @extern(
+        *const fn (
+            hProcess: os.windows.HANDLE,
+            dwMinimumWorkingSetSize: os.windows.SIZE_T,
+            dwMaximumWorkingSetSize: os.windows.SIZE_T,
+        ) callconv(os.windows.WINAPI) os.windows.BOOL,
+        .{
+            .library_name = "kernel32",
+            .name = "SetProcessWorkingSetSize",
+        },
+    );
+    const get_process_working_set_size = @extern(
+        *const fn (
+            hProcess: os.windows.HANDLE,
+            lpMinimumWorkingSetSize: *os.windows.SIZE_T,
+            lpMaximumWorkingSetSize: *os.windows.SIZE_T,
+        ) callconv(os.windows.WINAPI) os.windows.BOOL,
+        .{
+            .library_name = "kernel32",
+            .name = "GetProcessWorkingSetSize",
+        },
+    );
+
+    const process_handle = os.windows.kernel32.GetCurrentProcess();
+    var working_set_min: os.windows.SIZE_T = 0;
+    var working_set_max: os.windows.SIZE_T = 0;
+
+    if (get_process_working_set_size(
+        process_handle,
+        &working_set_min,
+        &working_set_max,
+    ) == os.windows.FALSE) {
+        working_set_min = allocated_size; // Count bytes allocated so far.
+        working_set_min += 64 * 1024 * 1024; // 64mb buffer room for stack/globals.
+        working_set_max = working_set_min * 2; // Buffer room for new page faults.
```

Comment:
This feels suspicious --- does this doubles our memory size? And also, it feels rather wrong --- we _could_ have large fragmentation here, the relation between the size allocated and the number of pages we need is non-trivial...

But out of scope for this.